packages/protocol/contracts/L1/libs/LibProposing.sol

```diff
@@ -64,56 +68,116 @@ library LibProposing {
+        returns (TaikoData.BlockMetadata2 memory meta_, TaikoData.EthDeposit[] memory deposits_)
     {
-        Local memory local;
-        local.params = abi.decode(_data, (TaikoData.BlockParams));
-
-        if (local.params.coinbase == address(0)) {
-            local.params.coinbase = msg.sender;
+        // Checks proposer access.
+        {
+            address access = _resolver.resolve(LibStrings.B_PROPOSER_ACCESS, true);
+            if (access != address(0) && !IProposerAccess(access).isProposerEligible(msg.sender)) {
+                revert L1_INVALID_PROPOSER();
+            }
         }
 
         // Taiko, as a Based Rollup, enables permissionless block proposals.
+        Local memory local;
         local.b = _state.slotB;
+        local.postFork = local.b.numBlocks >= _config.ontakeForkHeight;
 
         // It's essential to ensure that the ring buffer for proposed blocks
         // still has space for at least one more block.
         if (local.b.numBlocks >= local.b.lastVerifiedBlockId + _config.blockMaxProposals + 1) {
             revert L1_TOO_MANY_BLOCKS();
         }
 
-        local.parentMetaHash =
-            _state.blocks[(local.b.numBlocks - 1) % _config.blockRingBufferSize].metaHash;
-        // assert(parentMetaHash != 0);
+        TaikoData.BlockParams2 memory params;
+
+        if (local.postFork) {
+            if (_data.length != 0) {
+                params = abi.decode(_data, (TaikoData.BlockParams2));
+                // otherwise use a default BlockParams2 with 0 values
+            }
+        } else {
+            params = LibData.paramV1toV2(abi.decode(_data, (TaikoData.BlockParams)));
+        }
+
+        if (params.coinbase == address(0)) {
+            params.coinbase = msg.sender;
+        }
+
+        if (!local.postFork || params.anchorBlockId == 0) {
+            params.anchorBlockId = uint64(block.number - 1);
+        }
+
+        if (!local.postFork || params.timestamp == 0) {
+            params.timestamp = uint64(block.timestamp);
+        }
 
-        // Check if parent block has the right meta hash. This is to allow the proposer to make sure
-        // the block builds on the expected latest chain state.
-        if (local.params.parentMetaHash != 0 && local.parentMetaHash != local.params.parentMetaHash)
+        // Verify params against the parent block.
         {
-            revert L1_UNEXPECTED_PARENT();
+            TaikoData.Block storage parentBlk =
+                _state.blocks[(local.b.numBlocks - 1) % _config.blockRingBufferSize];
+
+            if (local.postFork) {
+                // Verify the passed in L1 state block number.
+                // We only allow the L1 block to be 2 epochs old.
+                // The other constraint is that the L1 block number needs to be larger than or equal
+                // the
+                // one
+                // in the previous L2 block.
+                if (
+                    params.anchorBlockId + 64 < block.number //
```

Comment:
I think last info was to make it 4 epoch instead of 2. https://github.com/taikoxyz/taiko-mono/pull/17654#issuecomment-2196108578

Reply:
I introduced a config var so testnets can have a different value.